lib/src/data/settings/settings_local_repository.dart

```diff
@@ -0,0 +1,38 @@
+import 'package:allthenews/src/domain/common/persistence/persistence_repository.dart';
+import 'package:allthenews/src/domain/settings/popular_news_criterion.dart';
+import 'package:allthenews/src/domain/settings/settings_repository.dart';
+import 'package:allthenews/src/domain/settings/app_theme.dart';
+
+class _Constants {
+  static const selectedPopularNewsCriterionKey = 'popularNewsCriterionKey';
+  static const selectedThemeKey = 'selectedThemeKey';
+}
+
+class SettingsLocalRepository extends SettingsRepository {
+
+  final PersistenceRepository _persistenceRepository;
+
+  SettingsLocalRepository(this._persistenceRepository);
+
+  @override
+  Future<PopularNewsCriterion> getPopularNewsCriterion() async {
+    final criterionValue = await _persistenceRepository.getString(_Constants.selectedPopularNewsCriterionKey) ?? PopularNewsCriterion.viewed.toString();
+    return PopularNewsCriterion.values.firstWhere((criterion) => criterion.toString() == criterionValue, orElse: () => PopularNewsCriterion.viewed);
+  }
+
+  @override
+  Future<AppTheme> getTheme() async {
+    final themeValue = await _persistenceRepository.getString(_Constants.selectedThemeKey) ?? AppTheme.light.toString();
+    return AppTheme.values.firstWhere((theme) => theme.toString() == themeValue, orElse: () => AppTheme.light);
+  }
+
+  @override
+  Future<void> savePopularNewsCriterion(PopularNewsCriterion popularNewsCriterion) async {
```

Comment:
enuma imo lepiej przechowywać jako inta w prefsach

Reply:
czemu? ;) (chyba nie ze względów wydajnościowych?) - ta wartość miałaby bazować na indeksie czy widziałbyś to jako metody mapujące int -> enum i enum -> int?

Reply:
int wydaje mi się bardziej naturalny dla enuma, właśnie dlatego, że ma index i mapowanie na stringa wydaje się nadmiarowe. Zapis do prefsów będzie podobny, ale wyciąganie będzie prostsze 
zamiast
`    return PopularNewsCriterion.values.firstWhere((criterion) => criterion.index == criterionValue, orElse: () => PopularNewsCriterion.viewed);
`
będzie 
`return PopularNewsCriterion.values[criterionValue];`

Reply:
Tak chwilę pomyślałem jakie są wady i zalety obu podejść i w przypadku przechowywania stringa to zaletą jest czytelność, generalnie nie ma problemów z dodawaniem, usuwaniem i kolejnością wartości enuma - a także z tego co widzę to chyba nie można nadpisać toStringa w enumie? wadą natomiast to, że polegasz na toStringu i że jak zmienisz nazwę no efektem będzie to, że nie znajdzie takiego elementu i wrzuci domyślny. W przypadku inta zaletami jest minimalnie lepsza wydajność, minimalnie łatwiejsze wyciąganie i nawet jak się zmieni nazwę, a zachowa kolejność to będzie ok. Wadami jest natomiast brak czytelności ('Theme.light' > 0), nieodporność na zmianę kolejności i na usuwanie elementów enuma - w przypadku zmiany kolejności zwrócona zostanie inna wartość niż zapisana, a w przypadku usunięcia i wyjścia poza zakres tablicy wywali RangeError (tu można się zabezpieczyć, ale wtedy wyciąganie nie będzie takie ładne). Nie wiem czy wszystko uwzględniłem, żadne podejście nie wydaje się idealne, ale wydaje mi się, że chyba string jest mniej podatny na błędy?

Reply:
wiesz, wiele rzeczy jakbyśmy zmienili to byśmy mieli wyrzucony wyjątek (na przykład wartości kluczy, kolejność innych enumów) i jeżeli mamy takie obawy to możemy zacząć pisać testy.
Co do czytelności to nie widzę nigdzie, żeby była konieczność porównywania gołego Stringa / inta z enumem, ale nawet jeśli to przecież nie trzeba robić tego poprzez `Theme.light > 0' tylko można odnieść się do wartości na indeksie i wtedy będzie jeszcze czytelniej niż rzutowanie enuma do stringa, nie wiem gdzie widać tu przewagę czytelności stringa nad intem?

W rozwiązaniu ze stringami najbardziej mnie boli to, że potrzebna jest logika do wyciągania wartości z enuma i podwójnie musimy definiować domyślną wartość. Umniejszyłeś to do `minimalnie łatwiejszego wyciągania`

Reply:
Hmm... w zasadzie to nie trzeba definiować 2 razy domyślnej wartości, już poprawiłem ;) Co do czytelności to nieprecyzyjnie napisałem. Theme.light > 0 - chodziło mi o to że wartość 'Theme.light' jest bardziej czytelna niż '0' - debugujesz sobie apkę, patrzysz co masz w shared prefach, widzisz theme: 0 - no to jest mniej czytelne niż Theme.light. Tak samo przy ewentualnych problemach łatwiej jest przeanalizować widząc formę bardziej czytelną dla człowieka. W przypadku zmiany wartości klucza jeżeli wyciąganie zakłada, że może tam nic nie być to nie ma problemu, bo po prostu nic nie znajdzie, jedyne co, to jeżeli zmienisz klucz przechowywanego inta na klucz czegoś co przechowuje stringa, ale takie przypadki chyba się nie zdarzają :P A przykład z innymi enumami, zgadzam się, ale właśnie dlatego uważam poleganie na indeksie za śliskie. Żeby przetestować taki scenariusz musiałyby być to testy sprawdzające update aplikacji z jednej wersji do drugiej czy zachowana jest kompatybilność, co najłatwiejsze nie jest ;)

Reply:
dlatego trochę zaskoczył mnie argument, że kolejność enuma się może zmienić i w tym przewaga logiki wyszukiwania stringa, bo wiele rzeczy się wysypie jeżeli coś się zmieni

Reply:
https://stackoverflow.com/a/22951509 tu jest dokładnie ten case, o którym mówimy

Reply:
teraz jak się pozbyliśmy podwójnego definiowania default value to już nie wygląda tak źle :)

Reply:
też się naszukałem oficjalnych guideline/best practices ale na próżno w darcie/flutterze :P